File: src/providers/status.ts
import type { CodexStatus, StatusField, SyncState } from "./types";

const blank = <T>(): StatusField<T> => ({ value: null, source: null, observedAt: null, confidence: null, stale: false });

export function isFailureSyncState(state: SyncState): boolean {
  return ["disconnected", "stale", "recon-required", "unsupported"].includes(state);
}

export function emptyStatus(message: string | null = null): CodexStatus {
  return {
    username: blank(), model: blank(), reasoningEffort: blank(), reasoningSpeed: blank(), speedMode: blank(),
    subscription: blank(), remainingPercent: blank(), resetAt: blank(), clientVersion: blank(),
    monthlyUsage: blank(), weeklyDurationSeconds: blank(),
    syncState: message ? "recon-required" : "idle", message
  };
}

export function mergeStatus(previous: CodexStatus, incoming: CodexStatus): CodexStatus {
  const keys = ["username", "model", "reasoningEffort", "reasoningSpeed", "speedMode", "subscription", "remainingPercent", "resetAt", "clientVersion", "monthlyUsage", "weeklyDurationSeconds"] as const;
  const merged = { ...incoming };
  for (const key of keys) {
    if (incoming[key].value === null && previous[key].value !== null) merged[key] = { ...previous[key], stale: true } as never;
  }
  return merged;
}

File: src-tauri/src/status.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusField<T> { pub value: Option<T>, pub source: Option<String>, pub observed_at: Option<String>, pub confidence: Option<f32>, pub stale: bool }
impl<T> Default for StatusField<T> { fn default() -> Self { Self { value: None, source: None, observed_at: None, confidence: None, stale: false } } }

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexStatus {
    pub username: StatusField<String>, pub model: StatusField<String>, pub reasoning_effort: StatusField<String>,
    pub reasoning_speed: StatusField<f64>, pub speed_mode: StatusField<String>, pub subscription: StatusField<String>, pub remaining_percent: StatusField<f64>,
    pub reset_at: StatusField<String>, pub client_version: StatusField<String>, pub monthly_usage: StatusField<f64>,
    pub weekly_duration_seconds: StatusField<f64>, pub sync_state: String, pub message: Option<String>,
}
impl CodexStatus {
    pub fn unavailable(message: impl Into<String>) -> Self { Self { sync_state: "recon-required".into(), message: Some(message.into()), ..Self::default() } }
    pub fn unsupported(message: impl Into<String>) -> Self { Self { sync_state: "unsupported".into(), message: Some(message.into()), ..Self::default() } }
}
